packages/headless/src/controllers/product-listing/sort/headless-product-listing-sort.ts

```diff
@@ -0,0 +1,66 @@
+import {SortCriterion} from '../../../features/sort-criteria/criteria';
+import {
+  buildCoreSort,
+  Sort,
+  SortInitialState,
+  SortProps,
+  SortState,
+} from '../../core/sort/headless-core-sort';
+import {fetchProductListing} from '../../../features/product-listing/product-listing-actions';
+import {ProductListingEngine} from '../../../app/product-listing-engine/product-listing-engine';
+import {parseCriterionExpression} from '../../../features/sort-criteria/criteria-parser';
+
+export interface ProductListingSortState extends SortState {
+  /**
+   * The current sort criteria, in a structured way.
+   */
+  sort: SortCriterion | SortCriterion[];
+}
+
+export interface ProductListingSort extends Sort {
+  /**
+   * A scoped and simplified part of the headless state that is relevant to the `Sort` controller.
+   * */
+  state: ProductListingSortState;
+}
+
+export {
+  SortInitialState,
+  SortProps,
+  ProductListingSort as Sort,
+  ProductListingSortState as SortState,
+};
+
+/**
+ * Creates a `Sort` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `Sort` controller properties.
+ * @returns A `Sort` controller instance.
+ */
+export function buildSort(
+  engine: ProductListingEngine,
+  props: SortProps = {}
+): ProductListingSort {
+  const {dispatch} = engine;
+  const sort = buildCoreSort(engine, props);
+  const getState = () => engine.state;
+
+  return {
+    ...sort,
+
+    sortBy(criterion: SortCriterion | SortCriterion[]) {
+      sort.sortBy(criterion);
+      dispatch(fetchProductListing());
+    },
+
+    get state() {
+      return {
+        ...sort.state,
+        sort: getState().sortCriteria
+          ? parseCriterionExpression(getState().sortCriteria!)
```

Comment:
Hmm, some thoughts:

I'd expect `getState().sortCriteria` to be truthy since the slice was registered by core reducer. Why is a ternary needed? Could we get the value from `sort.state.sortCriteria`?

Could the key in state be called `sortCriteria` for consistency? With the destructuring, it seems like there would be two keys which is not as nice?

It would also be good to add a unit test for the parsing _if_ it is needed. Otherwise, maybe exposing a string is perfectly fine?

Reply:
> I'd expect getState().sortCriteria to be truthy since the slice was registered by core reducer. Why is a ternary needed? Could we get the value from sort.state.sortCriteria?

Yes! I'll make the change.

> Could the key in state be called sortCriteria for consistency? With the destructuring, it seems like there would be two keys which is not as nice?

If we rename the `sort` key in the state to `sortCriteria`, it'll clash with the other `sortCriteria` key which is already present (in `...sort.state`), so I don't think we can.

> It would also be good to add a unit test for the parsing if it is needed. Otherwise, maybe exposing a string is perfectly fine?

I'll add a test case to make sure `parseCriterionExpression` is called with the `sortCriteria` to create the `sort` key.